homeassistant/components/tuya/__init__.py

```diff
@@ -107,86 +122,154 @@ def load_devices(device_list):
+                )
+                hass.data[CONFIG_ENTRY_IS_SETUP].add(config_entries_key)
+            else:
+                async_dispatcher_send(hass, TUYA_DISCOVERY_NEW.format(ha_type), dev_ids)
+
+    device_list = await hass.async_add_executor_job(tuya.get_all_devices)
+    await async_load_devices(device_list)
+
+    async def async_poll_devices_update(event_time):
         """Check if accesstoken is expired and pull device list from server."""
         _LOGGER.debug("Pull devices from Tuya.")
-        tuya.poll_devices_update()
+        await hass.async_add_executor_job(tuya.poll_devices_update)
         # Add new discover device.
-        device_list = tuya.get_all_devices()
-        load_devices(device_list)
+        device_list = await hass.async_add_executor_job(tuya.get_all_devices)
+        await async_load_devices(device_list)
         # Delete not exist device.
         newlist_ids = []
         for device in device_list:
             newlist_ids.append(device.object_id())
         for dev_id in list(hass.data[DOMAIN]["entities"]):
             if dev_id not in newlist_ids:
-                dispatcher_send(hass, SIGNAL_DELETE_ENTITY, dev_id)
+                async_dispatcher_send(hass, SIGNAL_DELETE_ENTITY, dev_id)
                 hass.data[DOMAIN]["entities"].pop(dev_id)
 
-    track_time_interval(hass, poll_devices_update, timedelta(minutes=5))
+    hass.data[TUYA_TRACKER] = async_track_time_interval(
+        hass, async_poll_devices_update, timedelta(minutes=5)
+    )
 
-    hass.services.register(DOMAIN, SERVICE_PULL_DEVICES, poll_devices_update)
+    hass.services.async_register(
+        DOMAIN, SERVICE_PULL_DEVICES, async_poll_devices_update
+    )
 
-    def force_update(call):
+    async def async_force_update(call):
         """Force all devices to pull data."""
-        dispatcher_send(hass, SIGNAL_UPDATE_ENTITY)
+        async_dispatcher_send(hass, SIGNAL_UPDATE_ENTITY)
 
-    hass.services.register(DOMAIN, SERVICE_FORCE_UPDATE, force_update)
+    hass.services.async_register(DOMAIN, SERVICE_FORCE_UPDATE, async_force_update)
 
     return True
 
 
+async def async_unload_entry(hass, entry):
+    """Unloading the Tuya platforms."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(
+                    entry, component.split(".", 1)[0]
+                )
+                for component in hass.data[CONFIG_ENTRY_IS_SETUP]
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[CONFIG_ENTRY_IS_SETUP] = set()
+        hass.data[TUYA_TRACKER]()
+        hass.data[TUYA_TRACKER] = None
+        hass.data[TUYA_DATA] = None
+        hass.services.async_remove(DOMAIN, SERVICE_FORCE_UPDATE)
+        hass.services.async_remove(DOMAIN, SERVICE_PULL_DEVICES)
+        hass.data.pop(DOMAIN)
+
+    return unload_ok
+
+
+async def cleanup_device_registry(hass, device_id):
```

Comment:
Please check if this is still needed as it may be done automatically since #35106 is now merged

Reply:
Here I need to rebase, I will check tomorrow.

Reply:
No rebase needed, PR was already there (too late yesterday).
I confirm that device is removed automatically and I removed this part of code. 
Test are failing due to internal error (flake unavailable), let me know when I can force test again.